integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/AsyncCompositeHealthIndicator.java

```diff
@@ -0,0 +1,167 @@
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link HealthIndicator} that returns health indications from all registered delegates using
+ * asynchronous calls.
+ * Created by rsanchez on 16/01/17.
+ */
+public class AsyncCompositeHealthIndicator implements HealthIndicator {
+
+  /**
+   * Thread pool size
+   */
+  private static final Integer MAX_POOL_SIZE = 10;
+
+  /**
+   * Timeout to verify if the execution was done
+   */
+  private static final Long EXECUTION_TIMEOUT = 1000L;
+
+  /**
+   * Registered indicators
+   */
+  private final Map<String, HealthIndicator> indicators;
+
+  /**
+   * Health aggregator
+   */
+  private final HealthAggregator healthAggregator;
+
+  public AsyncCompositeHealthIndicator(HealthAggregator aggregator) {
+    this.healthAggregator = aggregator;
+    this.indicators = new HashMap<>();
+  }
+
+  /**
+   * Register new indicator
+   * @param name Health indicator name
+   * @param indicator Health indicator object
+   */
+  public void addHealthIndicator(String name, HealthIndicator indicator) {
+    this.indicators.put(name, indicator);
+  }
+
+  @Override
+  public Health health() {
+    try {
+      Map<String, Future<Health>> result = asyncExecution();
+      Map<String, Health> healths = extractResult(result);
+      return this.healthAggregator.aggregate(healths);
+    } catch (InterruptedException e) {
+      return Health.down().withException(e).build();
+    }
+  }
+
+  /**
+   * Executes registered indicators using asynchronous calls.
+   * @return
+   * @throws InterruptedException
+   */
+  private Map<String, Future<Health>> asyncExecution() throws InterruptedException {
+    Map<String, Future<Health>> result = new LinkedHashMap<>();
+
+    int indicatorSize = Math.max(1, indicators.size());
+    int poolSize = Math.min(MAX_POOL_SIZE, indicatorSize);
+    ExecutorService service = Executors.newFixedThreadPool(poolSize);
+
+    for (Map.Entry<String, HealthIndicator> entry : indicators.entrySet()) {
+      final HealthIndicator indicator = entry.getValue();
+
+      Future<Health> execution = service.submit(new Callable<Health>() {
+        @Override
+        public Health call() throws Exception {
+          return indicator.health();
+        }
+      });
+
+      result.put(entry.getKey(), execution);
+    }
+
+    shutdownService(service);
+
+    return result;
+  }
+
+  /**
+   * Shutdown the executor service
+   * @param service Executor Service object
+   * @throws InterruptedException Thread execution interrupted
+   */
+  private void shutdownService(ExecutorService service) throws InterruptedException {
+    service.shutdown();
+
+    for (int i = 0; i < 10; i++) {
```

Comment:
Be more descriptive, instead of "i" and fixed numbers use constants and more relevant names here.